src/functions.rs

```diff
@@ -2024,10 +2103,32 @@ impl ValidatedCall {
             .into());
         }
 
-        // Each argument must match the type the function signature expects
-        let mut args = match &call.class_name {
-            Some(c) => vec![convert_class_name_if_this(c, parent_type)?.get_cil_name()],
-            None => Vec::new(),
+        if let Some(cast_name) = &call.cast_name {
+            if let Some(class_name) = &call.class_name {
+                validate_cast(
+                    class_name,
+                    types.get(cast_name.as_ref()),
+                    Some(call),
+                    types,
+                    context,
+                    Some(file),
+                )?;
+            }
+        }
+
+        // All unwraps here are safe due to if checks
+        let mut args = if call.class_name.is_none() && call.cast_name.is_some() {
```

Comment:
You should be able to write this more cleanly as a match statement I think.  Something like:

```
let mut args = match (call.class_name, call.cast_name) {
    (None, Some(cast_name)) => { ... },
    (None, None) => { ... },
    (Some(class_name), _) => { ... },
}
```

Reply:
Of course, rewriting it that way really suggests a further rewrite to:

```
let mut args = match call.class_name {
    Some(class_name) => { ... },
    None => { match call.cast_name {
        Some(cast_name) => { ... },
        None => { ... },
    }
}
```

which might make the intent cleaner.  I'm not sure I've actually followed the logic of this part yet though.